### ISSUE TITLE
Terminal left in semi-unusable state after `gatsby build`

### ISSUE THREAD

**Issue by tony-o** [gatsbyjs/gatsby#37512]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

The terminal is left in a state different than gatsby found it.  Running `gatsby build` results in the terminal no longer handling ctrl+d properly and eats the tab completion.  This leads to things like `:!gatsby build` in vim requiring you exit vim because the editor is no longer usable.  The terminal corrects itself when ctrl+c is issued and I think this may be related to #16621.  

This is what happens after running the `:!gatsby build` from vim:

![image](https://user-images.githubusercontent.com/2029968/213882725-4717db60-2b5d-4390-8662-3faf3e62b6a8.png)

### Reproduction Link

https://githubbox.com/gatsbyjs/gatsby-starter-minimal

### Steps to Reproduce

1. Run bash
2. Run `gatsby build`
3. Try to use tab completion
4. Hit Ctrl+C
5. Try to use tab completion

### Expected Result

3. Bash prompts you for commands it's aware of
5. Bash prompts you for commands it's aware of

### Actual Result

3. Tab characters are printed out
5. Bash prompts you for commands it's aware of

### Environment

```shell
System:
    OS: Linux 6.0 openSUSE Tumbleweed 20221209
    CPU: (40) x64 Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz
    Shell: 5.2.12 - /bin/bash
  Binaries:
    Node: 19.4.0 - ~/.local/src/nvm/versions/node/v19.4.0/bin/node
    Yarn: 1.22.19 - /usr/bin/yarn
    npm: 9.2.0 - ~/.local/src/nvm/versions/node/v19.4.0/bin/npm
  Languages:
    Python: 2.7.18 - /usr/bin/python
  Browsers:
    Chrome: 108.0.5359.98
    Firefox: 109.0
  npmPackages:
    gatsby: ^5.4.2 => 5.4.2
```

**Comment by pieh** [comment]
In effort to narrow things down a bit - could you try using `GATSBY_LOGGER=yurnalist` env var together with `gatsby build` and report wether it messes up terminal the same as default logger?

**Comment by tony-o** [comment]
@pieh just gave that a go and, yes, the logger env var+`gatsby build` still leaves the terminal in a strange state

**Comment by tony-o** [comment]
This is also happening with the same version of bash on OSX.  It's almost like the terminal is left with half of a control character sent, iterating search history via `ctrl+r` or `up arrow` and hitting enter still works but nothing prints to the console until the command itself prints and it prints in color as you'd expect.

I also tried building with the no color option set and that resulted in being left with the same unusuable vim/terminal.

Hitting `tab` a few times prints the tab characters, hitting `enter` directly afterwards prompts for command completion.  This led me to believe that whatever eats STDIN during the process was being left open after the process completes but I ran out of time troubleshooting but did notice that if I `gatsby build &` my terminal's cursor was gone for the rest of the session (even with several `ctrl+c`s).
